### PR TITLE
[MAINTENANCE] Refactoring of `test_script_runner.py`. Integration and Docs tests

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
Changes proposed in this pull request:
# Refactoring of `test_script_runner.py`
- This PR introduces the `IntegrationTestFixture` `dataclass` object.
- Configurations for integration tests are defined as lists of `IntegrationTestFixture` objects, which then create the environment, load the data, and run the tests, similar to how `test_script_runner.py` was written before. 
- `Name` field is required for `IntegrationTestFixtures` so tests can also be run by setting the '-k' flag and referencing the name of test:
    `pytest -v --docs-tests -m integration -k "test_docs[migration_guide_spark_v2_api]" `
     - **Note**: Before,  the tests were getting parameterized by the path of the script, which included `/` characters. This meant that pytest would give an error whenever we tried to name an individual test and run it. 
- Test scripts are now loaded and run as part of `module` using Python's `importlib` machinery. This means that tests are no longer run as a `subprocess` and errors and warnings are presented directly into the console.
- This brings us to.....
## A small war on warnings
- Running the test scripts as `modules` revealed a number of warnings in our tests that we simply weren't seeing before. These included: 

1. Invalid escape characters in REGEX (the `\` in `"yellow_tripdata_(\d{4})-(\d{2})\.csv",`). This was solved by turning the string into a `r` (raw) string. 

2. Missing installation of `google-cloud-bigquery-storage`.This was solved by adding it to `azure-pipelines-docs-integration.yml`. 

3. setting relative path in our test script like:
```
 path_to_file: str = "./data/yellow_tripdata_sample_2019-01.csv"
```
was causing sporadic issues with integration tests, especially when they were run in a series. Using a more explicit relative path:
```
path_to_file: str = file_relative_path(
    __file__, "data/yellow_tripdata_sample_2019-01.csv"
)
```
solved the issue and the remaining tests will be modified to reflect this pattern. 
 4.  Some methods were using `ruamel.yaml`'s `safe_load()` function, which will be deprecated. This is actually where most of our warnings are coming from. Going forward we will be using the pattern: 
```
yaml = yaml.YAML(typ="safe")
great_expectations_yaml = yaml.load(f)
```
- Importantly, this change affects the line numbers for every DOC that uses YAML configurations, which is 74 files. A future PR will implement the pattern suggested by `ruamel.yaml` .
- This additional work can be tracked in the branch : `https://github.com/great-expectations/great_expectations/tree/MAINTENANCE/GREAT-575/war-on-warnings-pytest-runner`

### Additional BugFixes / refactorings
- Util methods for `awsathena.py` were moved from `tests/integration/db/utils.py` (which was only being used by `awsathena` to `tests/test_utils.py`, the general utils files for tests. 
- `redshift` test was still referring to old `load_data_into_database` method. This was changed to `load_data_into_test_database`

Previous Design Review notes:
- GREAT-535
- enables GREAT-575

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: azure-pipelines-docs-integration.yml [hunk 1]
```diff
@@ -110,6 +110,8 @@ stages:
               secureFile: 'aws_config_sandbox'
               retryCount: '2'
         - script: |
+            # install google cloud storage dependency
+            pip install google-cloud-bigquery-storage
```
Comment (Shinnnyshinshin) [inline azure-pipelines-docs-integration.yml:114]: This was giving us a warning

**Review by alexsherstinsky — APPROVED** [review]
LGTM

File: tests/integration/docusaurus/connecting_to_your_data/how_to_create_a_batch_of_data_from_an_in_memory_spark_dataframe.py [hunk 2]
```diff
@@ -5,10 +5,13 @@
 from great_expectations import DataContext
 from great_expectations.core import ExpectationSuite
 from great_expectations.core.batch import RuntimeBatchRequest
+from great_expectations.data_context.util import file_relative_path
 from great_expectations.validator.validator import Validator
 
 context: DataContext = ge.get_context()
 
+yaml = yaml.YAML(typ="safe")
```
Comment (cdkini) [inline tests/integration/docusaurus/connecting_to_your_data/how_to_create_a_batch_of_data_from_an_in_memory_spark_dataframe.py:13]: kwarg typo?

File: tests/integration/docusaurus/connecting_to_your_data/how_to_create_a_batch_of_data_from_an_in_memory_spark_dataframe.py [hunk 3]
```diff
@@ -42,7 +45,9 @@
 path_to_file: str = "some_path.csv"
 
 # Please note this override is only to provide good UX for docs and tests.
-path_to_file: str = "./data/yellow_tripdata_sample_2019-01.csv"
+path_to_file: str = file_relative_path(
+    __file__, "data/yellow_tripdata_sample_2019-01.csv"
+)
```
Comment (cdkini) [inline tests/integration/docusaurus/connecting_to_your_data/how_to_create_a_batch_of_data_from_an_in_memory_spark_dataframe.py:50]: ❤️

File: tests/integration/docusaurus/miscellaneous/migration_guide_spark_v2_api.py [hunk 4]
```diff
@@ -3,20 +3,23 @@
 from ruamel import yaml
 
 import great_expectations as ge
+from great_expectations.data_context.util import file_relative_path
 
 context = ge.get_context()
 
+yaml = yaml.YAML(typ="safe")
```
Comment (cdkini) [inline tests/integration/docusaurus/miscellaneous/migration_guide_spark_v2_api.py:10]: Just kidding I looked at the official docs. Wow I do not love the naming convention from ruamel (but I love this change here!)

File: tests/integration/usage_statistics/test_usage_stats_common_messages_are_sent_v3api.py [hunk 5]
```diff
@@ -117,7 +117,7 @@ def test_common_usage_stats_are_sent_no_mocking(
     context.test_yaml_config(yaml_config=datasource_yaml)
     expected_events: List[str] = ["data_context.test_yaml_config"]
 
-    context.add_datasource(**yaml.load(datasource_yaml))
+    context.add_datasource(**yaml.safe_load(datasource_yaml))
```
Comment (cdkini) [inline tests/integration/usage_statistics/test_usage_stats_common_messages_are_sent_v3api.py]: Why is this flipped with L160?

File: tests/test_utils.py [hunk 6]
```diff
@@ -494,3 +494,59 @@ def set_directory(path: str) -> Generator:
         yield
     finally:
         os.chdir(origin)
+
+
+def check_athena_table_count(
+    connection_string: str, db_name: str, expected_table_count: int
+) -> bool:
```
Comment (cdkini) [inline tests/test_utils.py:501]: Lovely type annotations, thank you for this 🙇🏽

File: tests/integration/test_script_runner.py [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
-from typing import Tuple
+from dataclasses import dataclass
+from typing import List, Optional, Tuple
 
 import pytest
 
 from assets.scripts.build_gallery import execute_shell_command
 from great_expectations.data_context.util import file_relative_path
 
+logger = logging.getLogger(__name__)
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:16]: Should we set our default log level to something lower than the default (WARNING)?

File: tests/integration/test_script_runner.py [hunk 8]
```diff
@@ -25,476 +30,582 @@ class BackendDependencies(enum.Enum):
...
 
 
+@dataclass
+class IntegrationTestFixture:
+    """IntegrationTestFixture
+
+    Configurations for integration tests are defined as IntegrationTestFixture dataclass objects.
+
+    Individual tests can also be run by setting the '-k' flag and referencing the name of test, like the following example:
+    pytest -v --docs-tests -m integration -k "test_docs[migration_guide_spark_v2_api]" tests/integration/test_script_runner.py
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:41]: Woohoo 🎉

File: tests/integration/test_script_runner.py [hunk 9]
```diff
@@ -25,476 +30,582 @@ class BackendDependencies(enum.Enum):
...
+
+    Args:
+        name (str): Name for integration test. Individual tests can be run by using the -k option and specifying the name of the test.
+        user_flow_script (str): Required script for integration test.
+        data_context_dir Optional[str]: Path of great_expectations/ that is used in the test.
+        data_dir Optional[str]: Folder that contains data used in the test.
+        extra_backend_dependencies Optional[BackendDependencies]: Optional flag allows you to tie an individual test with a BackendDependency. Allows for tests to be run / disabled using cli flags (like --aws which enables AWS integration tests).
+        other_files Optional[Tuple[Tuple[str, str]]]: other files (like credential information) to copy into the test environment.
+        util_script Optional[str]: Path of optional util script that is used in test script (for loading test_specific methods like load_data_into_test_database())
+        ge_requirement Optional[str]: allows you to set a specific version of GE to install. Default is current version of GE that is contained in the base directory (pip install .)
```
Comment (cdkini) [inline tests/integration/test_script_runner.py]: Appreciate the docstring here! I don't think we need types if we're explicitly noting them above.

File: tests/integration/test_script_runner.py [hunk 10]
```diff
@@ -25,476 +30,582 @@ class BackendDependencies(enum.Enum):
...
+    """
+
+    name: str
+    user_flow_script: str
+    data_context_dir: Optional[str] = None
+    data_dir: Optional[str] = None
+    extra_backend_dependencies: Optional[BackendDependencies] = None
+    other_files: Optional[Tuple[Tuple[str, str]]] = None
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:58]: `other_files` has a very particular structure; can we add a note in the docs about source/dest pairs and how that works?

File: tests/integration/test_script_runner.py [hunk 11]
```diff
@@ -25,476 +30,582 @@ class BackendDependencies(enum.Enum):
...
+    util_script: Optional[str] = None
+    ge_requirement: Optional[str] = None
```
Comment (cdkini) [inline tests/integration/test_script_runner.py]: We never actually use this. Should we remove it?

File: tests/integration/test_script_runner.py [hunk 12]
```diff
@@ -595,63 +703,54 @@ def test_integration_tests(test_configuration, tmp_path, pytest_parsed_arguments
     _execute_integration_test(test_configuration, tmp_path)
 
 
-def _execute_integration_test(test_configuration, tmp_path):
+def _execute_integration_test(integration_test_fixture, tmp_path):
     """
     Prepare and environment and run integration tests from a list of tests.
 
-    Note that the only required parameter for a test in the matrix is
+    N ote that the only required parameter for a test in the matrix is
```
Comment (cdkini) [inline tests/integration/test_script_runner.py]: Could we fix this back please?

File: tests/integration/test_script_runner.py [hunk 13]
```diff
@@ -595,63 +703,54 @@ def test_integration_tests(test_configuration, tmp_path, pytest_parsed_arguments
...
     """
-    assert (
-        "user_flow_script" in test_configuration.keys()
-    ), "a `user_flow_script` is required"
     workdir = os.getcwd()
     try:
-        base_dir = test_configuration.get(
-            "base_dir", file_relative_path(__file__, "../../")
-        )
+        base_dir = file_relative_path(__file__, "../../")
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:714]: Base dir is always the same so we removed it from the test config?

File: tests/integration/test_script_runner.py [hunk 14]
```diff
@@ -595,63 +703,54 @@ def test_integration_tests(test_configuration, tmp_path, pytest_parsed_arguments
...
-        if other_files is not None:
-            if not isinstance(other_files, Tuple):
-                raise TypeError("other_files must be of type Tuple[Tuple[str, str]]")
-            if not all(isinstance(t, Tuple) for t in other_files):
-                raise TypeError("other_files must be of type Tuple[Tuple[str, str]]")
-
+        other_files = integration_test_fixture.other_files
+        if other_files:
             for file_paths in other_files:
+                # check to see if these work
```
Comment (cdkini) [inline tests/integration/test_script_runner.py]: a note to self while developing?

File: tests/integration/test_script_runner.py [hunk 15]
```diff
@@ -660,55 +759,35 @@ def _execute_integration_test(test_configuration, tmp_path):
...
-            assert res.returncode == 0
-    except:
-        raise
+        # Run script as module, using python's importlib machinery (https://docs.python.org/3/library/importlib.htm)
+        loader = importlib.machinery.SourceFileLoader("test_script_module", script_path)
+        spec = importlib.util.spec_from_loader("test_script_module", loader)
+        test_script_module = importlib.util.module_from_spec(spec)
+        loader.exec_module(test_script_module)
+    except Exception as e:
+        logger.error(str(e))
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:776]: This is truly wonderful stuff! Very much appreciate the capturing of the Exception.

**Review by cdkini — APPROVED** [review]
Fantastic work! After all my complaining about the test script runner, it's a really nice feeling to have someone put in the work to really clean it up. The dataclasses add some much needed structure, the use of `importlib` is true wizardry, and the overall attention to detail is a sight to behold.

Left some very minor comments but nothing blocking!

File: tests/integration/test_script_runner.py (hunk 11 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py]: I think you gave me the nudge I needed to (finally) remove this. Consider it done :)

File: tests/integration/test_script_runner.py (hunk 12 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py]: 😓

File: tests/integration/test_script_runner.py (hunk 13 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py:714]: yes that's correct 👍🏼

File: tests/integration/test_script_runner.py (hunk 14 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py]: 😓

File: tests/integration/test_script_runner.py (hunk 10 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py:58]: Great suggestion. Adding now

**Pull request merged**